Indexer throws an exception when processing invalid spatial data from `Australian Oceans Data Network`

### Problem description

When testing my pr for #983, I found out indexer throws an exception when processing invalid spatial data from `Australian Oceans Data Network`.

This seems an existing issue on Master branch and currently stop indexer from going further.

You will see the error on Master Dev site (and any other namespaces) indexer pod log as well:

```
[ERROR] [12/07/2018 01:26:18.954] [indexer-akka.actor.default-dispatcher-991] [akka.actor.ActorSystemImpl(indexer)] Exception encountered
java.lang.NumberFormatException: multiple points
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1914)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
```
Create as a seperate ticket so that we can create a seperate PR for it. 

#### Root Cause

Around this line: 
https://github.com/magda-io/magda/blob/0ef9cb63117731ee5b2eca4852cd811e3bbdf676/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/Misc.scala#L140

We try to convert string like `123.8938` to a `double`. However, recent data from Australian Oceans Data Network has some data with more than one decimal places. i.e. `123.89.38`
Thus, produce the error above.

#### Proposed Solutions

- We should remove all `dots` after the first dot before call `_.double` to convert it. 
- Furthermore, considering low data quality in this area, we probably shouldn't let indexer stuck
  - The error was produced when a new webhook request is sent to indexer. Because of the error, this request will not be considered as `complete`. And you will see this log keep going and registry will not skip it to deliver the next one
  - Instead, we should:
    - Capture the error (around here: https://github.com/magda-io/magda/blob/0ef9cb63117731ee5b2eca4852cd811e3bbdf676/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/Registry.scala#L265)
    - Print an error log
    - Index the dataset without spatial data (set spatial data field to `None`)

#### Example Dataset
Dateset ID:  `ds-aodn-9f8e77e3-8ea8-44cb-96b5-3cd2efb38e51`
Invalid Spatial data: 
```
POLYGON((141.3241 -46.9723, 147.53.08 -46.9723, 147.53.08 -43.0053, 141.3241 -43.0053, 141.3241 -46.9723))
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -4,6 +4,7 @@
 -   Fixed chart won't displayed correctly under IE11
 -   Fixed stream processing issue when re-indexing, mitigating out-of-memory risk too
 -   Fixed index trimming failure issue when re-indexing
+-   Fixed: indexer throws an error when processes spatial data number with more than one decimal places
 
 ## 0.0.50
 

--- a/magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/external/registry/WebhookApi.scala
+++ b/magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/external/registry/WebhookApi.scala
@@ -27,7 +27,7 @@ class WebhookApi(indexer: SearchIndexer)(implicit system: ActorSystem, config: C
   /**
   * @apiGroup Indexer
   * @api {post} http://indexer/v0/registry-hook Hook endpoint (internal)
-  * 
+  *
   * @apiDescription Registry webhook endpoint - accepts webhook payloads from the registry.
   *   This generally means datasets from the registry as they're updated. This shouldn't
   *   be called manually, it's purely for registry use.
@@ -62,13 +62,13 @@ class WebhookApi(indexer: SearchIndexer)(implicit system: ActorSystem, config: C
             case None | Some(Nil) => Future.successful(Unit)
             case Some(list) =>
               val dataSets = list.map(record => try {
-                Some(convertRegistryDataSet(record))
+                Some(convertRegistryDataSet(record, Some(system.log)))
               } catch {
                 case CausedBy(e: spray.json.DeserializationException) =>
                   system.log.error(e, "When converting {}", record.id)
                   None
               })
-              
+
               indexer.index(Source(dataSets.flatten))
           }
 

--- a/magda-int-test/src/test/scala/au/csiro/data61/magda/indexer/WebhookSpec.scala
+++ b/magda-int-test/src/test/scala/au/csiro/data61/magda/indexer/WebhookSpec.scala
@@ -46,6 +46,12 @@ class WebhookSpec extends BaseApiSpec with RegistryConverters with ModelProtocol
 
   blockUntilNotRed()
 
+  def convertSpatialDataUsingGeoJsonField(spatialData:Option[Location]) = spatialData match {
+    case Some(Location(_, None)) => None
+    case Some(Location(_, Some(geoJsonData))) => Some(Location(Some(geoJsonData.toJson.toString)))
+    case _ => None
+  }
+
   describe("when webhook received") {
     it("should index new datasets") {
       loadDatasetsThroughEvents() { (allDataSets: List[DataSet], response: SearchResult) =>
@@ -77,12 +83,8 @@ class WebhookSpec extends BaseApiSpec with RegistryConverters with ModelProtocol
           // Contact points only look for name at the moment
           contactPoint = dataSet.contactPoint.flatMap(_.name).map(name => Agent(Some(name))),
 
-          // Registry doesn't know how to do spatial extent yet, so just keep the text, no text == None
-          spatial = dataSet.spatial match {
-            case Some(Location(None, _)) => None
-            case Some(spatial)           => Some(Location(spatial.text))
-            case other                   => other
-          },
+          // Registry doesn't know how to do spatial extent yet, so just keep the text and produce text from geoJson
+          spatial = convertSpatialDataUsingGeoJsonField(dataSet.spatial),
 
           temporal = dataSet.temporal match {
             // The registry -> dataset converter rightly does this conversion.
@@ -101,6 +103,7 @@ class WebhookSpec extends BaseApiSpec with RegistryConverters with ModelProtocol
           indexed = None,
           quality = 0,
           score = None,
+          spatial = convertSpatialDataUsingGeoJsonField(dataSet.spatial),
 
           distributions = dataSet.distributions.map(distribution => distribution.copy(
             // This will be derived from the format so might not match the input

--- a/magda-int-test/src/test/scala/au/csiro/data61/magda/test/util/Generators.scala
+++ b/magda-int-test/src/test/scala/au/csiro/data61/magda/test/util/Generators.scala
@@ -283,9 +283,22 @@ object Generators {
     }
 
   def locationGen(geometryGen: Gen[Geometry]) = for {
-    text <- someBiasedOption(textGen)
     geoJson <- someBiasedOption(geometryGen)
-  } yield new Location(text, geoJson)
+    shouldGenerateInvalidDecimals <- arbitrary[Boolean]
+  } yield {
+    val text = geoJson.map(_.toJson.toString).map { jsonString =>
+      if (!shouldGenerateInvalidDecimals) {
+        jsonString
+      } else {
+        /**
+          * Add extra invalid decimal places to simulate the possible poor geoJson data
+          * e.g. 123.34.22
+          * */
+        jsonString.replaceAll("(\\d+\\.\\d)(\\d+)", "$1.$2")
+      }
+    }
+    new Location(text, geoJson)
+  }
 
   def descWordGen(inputCache: mutable.Map[String, List[_]]) = cachedListGen("descWord", nonEmptyTextGen.map(_.take(50).mkString.trim), 200)(inputCache)
   def publisherAgentGen(inputCache: mutable.Map[String, List[_]]) =

--- a/magda-int-test/src/test/scala/au/csiro/data61/magda/test/util/Generators.scala
+++ b/magda-int-test/src/test/scala/au/csiro/data61/magda/test/util/Generators.scala
@@ -284,10 +284,30 @@ object Generators {
 
   def locationGen(geometryGen: Gen[Geometry]) = for {
     geoJson <- someBiasedOption(geometryGen)
+    shouldGenerateWKTString <- arbitrary[Boolean]
     shouldGenerateInvalidDecimals <- arbitrary[Boolean]
   } yield {
-    val text = geoJson.map(_.toJson.toString)
-      .map { jsonString =>
+    val text = geoJson.map { geoJsonObj =>
+      if (shouldGenerateWKTString && geoJsonObj.isInstanceOf[Polygon] && geoJsonObj.asInstanceOf[Polygon].coordinates.size < 2) {
+        // --- we only support WKT format for `Polygon`
+        // --- convert geoJson to WKT format
+        // --- We do not support Polygon in WKT format with holes yet
+        // --- Thus, only generate WKT text when generated Polygon has no holes
+        // --- i.e geoJsonObj.asInstanceOf[Polygon].coordinates.size < 2
+        val cordsStr = geoJsonObj
+                        .asInstanceOf[Polygon]
+                        .coordinates
+                        .toJson
+                        .toString
+                        .replaceAll("\\[([\\d-.]+).([\\d-.]+)\\]", "$1 $2")
+                        .replace("[", "(")
+                        .replace("]", ")")
+
+        s"POLYGON ${cordsStr}"
+      } else {
+        geoJsonObj.toJson.toString
+      }
+    }.map { jsonString =>
         if (!shouldGenerateInvalidDecimals) {
           jsonString
         } else {
@@ -298,6 +318,7 @@ object Generators {
           jsonString
             // --- add extra decimal places
             // --- please note: this will not be valid json anymore
+            // --- But still a valid WKT String (as per our pattern match)
             .replaceAll("(\\d+\\.\\d)(\\d+)", "$1.$2")
         }
     }

--- a/magda-int-test/src/test/scala/au/csiro/data61/magda/test/util/Generators.scala
+++ b/magda-int-test/src/test/scala/au/csiro/data61/magda/test/util/Generators.scala
@@ -286,16 +286,20 @@ object Generators {
     geoJson <- someBiasedOption(geometryGen)
     shouldGenerateInvalidDecimals <- arbitrary[Boolean]
   } yield {
-    val text = geoJson.map(_.toJson.toString).map { jsonString =>
-      if (!shouldGenerateInvalidDecimals) {
-        jsonString
-      } else {
-        /**
-          * Add extra invalid decimal places to simulate the possible poor geoJson data
-          * e.g. 123.34.22
-          * */
-        jsonString.replaceAll("(\\d+\\.\\d)(\\d+)", "$1.$2")
-      }
+    val text = geoJson.map(_.toJson.toString)
+      .map { jsonString =>
+        if (!shouldGenerateInvalidDecimals) {
+          jsonString
+        } else {
+          /**
+            * Add extra invalid decimal places to simulate the possible poor geoJson data
+            * e.g. 123.34.22
+            * */
+          jsonString
+            // --- add extra decimal places
+            // --- please note: this will not be valid json anymore
+            .replaceAll("(\\d+\\.\\d)(\\d+)", "$1.$2")
+        }
     }
     new Location(text, geoJson)
   }

--- a/magda-scala-common/src/main/scala/au/csiro/data61/magda/client/RegistryExternalInterface.scala
+++ b/magda-scala-common/src/main/scala/au/csiro/data61/magda/client/RegistryExternalInterface.scala
@@ -61,7 +61,7 @@ class RegistryExternalInterface(httpFetcher: HttpFetcher)
       response.status match {
         case OK => Unmarshal(response.entity).to[RegistryRecordsResponse].map { registryResponse =>
           (registryResponse.nextPageToken, mapCatching[Record, DataSet](registryResponse.records,
-            { hit => convertRegistryDataSet(hit) },
+            { hit => convertRegistryDataSet(hit, Some(logger)) },
             { (e, item) => logger.error(e, "Could not parse item: {}", item.toString) }))
         }
         case _ => Unmarshal(response.entity).to[String].flatMap(onError(response))
@@ -74,7 +74,7 @@ class RegistryExternalInterface(httpFetcher: HttpFetcher)
       response.status match {
         case OK => Unmarshal(response.entity).to[RegistryRecordsResponse].map { registryResponse =>
           (registryResponse.nextPageToken, mapCatching[Record, DataSet](registryResponse.records,
-            { hit => convertRegistryDataSet(hit) },
+            { hit => convertRegistryDataSet(hit, Some(logger)) },
             { (e, item) => logger.error(e, "Could not parse item: {}", item.toString) }))
         }
         case _ => Unmarshal(response.entity).to[String].flatMap(onError(response))

--- a/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/GeoJsonFormats.scala
+++ b/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/GeoJsonFormats.scala
@@ -5,7 +5,7 @@ import spray.json.DefaultJsonProtocol
 
 /**
  * Spray json marshallers for GeoJSON spec: http://geojson.org/geojson-spec.html
- * 
+ *
  * This is copy-pasted from mwundo because it hits a version conflict with spray json and causes the JVM to crash.
  */
 // scalastyle:off number.of.types
@@ -14,6 +14,32 @@ object GeoJsonFormats extends DefaultJsonProtocol {
   import spray.json._
 
   implicit object CoordinateFormat extends JsonFormat[Coordinate] {
+
+    private def cleanNumberString(str:String):String = {
+      val cleanedStr = str.replaceAll("\"","").trim
+      val parts = cleanedStr.split("\\.")
+      if (parts.length > 2) {
+        // --- drop all dots except the first one
+        parts.take(2).mkString(".") + parts.drop(2).mkString
+      } else {
+        cleanedStr
+      }
+    }
+
+    /**
+      * We sometimes call json.parse to convert the geoJson string to a json
+      * This will throw an exception for json data with invalid number data (e.g.  123.23)
+      * Quote all number in Json will avoid this exception and we will have the chance to convert it to
+      * BigDecimal
+      * */
+    def quoteNumbersInJson(str:String):String = {
+      str.replaceAll("([\\d-+.]+)", "\"$1\"")
+    }
+
+    def convertStringToBigDecimal(str:String):BigDecimal = {
+      BigDecimal(cleanNumberString(str))
+    }
+
     def write(obj: Coordinate): JsValue = JsArray(
       JsNumber(obj.x),
       JsNumber(obj.y)
@@ -21,7 +47,7 @@ object GeoJsonFormats extends DefaultJsonProtocol {
 
     def read(json: JsValue): Coordinate = json match {
       case JsArray(is) =>
-        Coordinate(is(0).convertTo[BigDecimal], is(1).convertTo[BigDecimal])
+        Coordinate(convertStringToBigDecimal(is(0).toJson.toString), convertStringToBigDecimal(is(1).toJson.toString))
       case _ => deserializationError(s"'$json' is not a valid Coordinate")
     }
   }

--- a/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/Misc.scala
+++ b/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/Misc.scala
@@ -10,6 +10,7 @@ import au.csiro.data61.magda.model.GeoJsonFormats._
 import au.csiro.data61.magda.model.Temporal._
 import spray.json._
 import scala.runtime.ScalaRunTime
+import scala.util.{ Failure, Success, Try }
 
 package misc {
   sealed trait FacetType {
@@ -128,25 +129,24 @@ package misc {
       val string = stringWithNewLines.replaceAll("[\\n\\r]", " ")
       Location.applySanitised(string, string match {
         case geoJsonPattern() => {
-          Some(Protocols.GeometryFormat.read(string.parseJson))
+          val json = Try(string.parseJson) match {
+            case Success(json) => json
+            case Failure(e) =>
+              CoordinateFormat.quoteNumbersInJson(string).parseJson
+          }
+          Some(Protocols.GeometryFormat.read(json))
         }
         case emptyPolygonPattern() => None
         case csvPattern(a) =>
-          val latLongs = string.split(",").map(BigDecimal.apply)
+          val latLongs = string.split(",").map(str => CoordinateFormat.convertStringToBigDecimal(str))
           fromBoundingBox(Seq(BoundingBox(latLongs(0), latLongs(1), latLongs(2), latLongs(3))))
         case polygonPattern(polygonCoords, _) =>
           val coords = polygonCoords.split(",")
             .map { stringCoords =>
-              val Array(x, y) = stringCoords.trim.split("\\s").map { str =>
-                // --- remove possible redundant dot in number string
-                val parts = str.split("\\.")
-                if (parts.length > 2) {
-                  // --- drop all dots except the first one
-                  parts.take(2).mkString(".") + parts.drop(2).mkString
-                } else {
-                  str
-                }
-              }.map(_.toDouble)
+              val Array(x, y) = stringCoords
+                                  .trim
+                                  .split("\\s")
+                                  .map(str => CoordinateFormat.convertStringToBigDecimal(str))
               Coordinate(x, y)
             }.toSeq
 

--- a/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/Misc.scala
+++ b/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/Misc.scala
@@ -137,7 +137,16 @@ package misc {
         case polygonPattern(polygonCoords, _) =>
           val coords = polygonCoords.split(",")
             .map { stringCoords =>
-              val Array(x, y) = stringCoords.trim.split("\\s").map(_.toDouble)
+              val Array(x, y) = stringCoords.trim.split("\\s").map { str =>
+                // --- remove possible redundant dot in number string
+                val parts = str.split("\\.")
+                if (parts.length > 2) {
+                  // --- drop all dots except the first one
+                  parts.take(2).mkString(".") + parts.drop(2).mkString
+                } else {
+                  str
+                }
+              }.map(_.toDouble)
               Coordinate(x, y)
             }.toSeq
 

--- a/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/Registry.scala
+++ b/magda-scala-common/src/main/scala/au/csiro/data61/magda/model/Registry.scala
@@ -16,8 +16,10 @@ import spray.json.DefaultJsonProtocol._
 import spray.json._
 import au.csiro.data61.magda.model.misc.{ Protocols => ModelProtocols }
 
-import scala.util.Try
+import scala.util.{ Failure, Success, Try }
 import java.time.format.DateTimeFormatter
+
+import akka.event.LoggingAdapter
 import au.csiro.data61.magda.util.DateParser
 
 object Registry {
@@ -216,7 +218,7 @@ object Registry {
       }
     }
 
-    def convertRegistryDataSet(hit: Record)(implicit defaultOffset: ZoneOffset): DataSet = {
+    def convertRegistryDataSet(hit: Record, logger: Option[LoggingAdapter] = None)(implicit defaultOffset: ZoneOffset): DataSet = {
       val dcatStrings = hit.aspects.getOrElse("dcat-dataset-strings", JsObject())
       val source = hit.aspects("source")
       val temporalCoverage = hit.aspects.getOrElse("temporal-coverage", JsObject())
@@ -252,6 +254,15 @@ object Registry {
         case (start, end)                           => Some(PeriodOfTime(Some(start), Some(end)))
       }
 
+      val spatialData = Try(dcatStrings.extract[String]('spatial.?).map(Location(_))) match {
+        case Success(location) => location
+        case Failure(e) =>
+          if(logger.isDefined) {
+            logger.get.error(s"Failed to parse spatial data for dataset ${hit.id}: ${e.getMessage}")
+          }
+          None
+      }
+
       DataSet(
         identifier = hit.id,
         title = dcatStrings.extract[String]('title.?),
@@ -262,7 +273,7 @@ object Registry {
         languages = dcatStrings.extract[String]('languages.? / *).toSet,
         publisher = publisher.map(convertPublisher),
         accrualPeriodicity = dcatStrings.extract[String]('accrualPeriodicity.?).map(Periodicity.fromString(_)),
-        spatial = dcatStrings.extract[String]('spatial.?).map(Location(_)), // TODO: move this to the CKAN Connector
+        spatial = spatialData, // TODO: move this to the CKAN Connector
         temporal = temporal,
         themes = dcatStrings.extract[String]('themes.? / *),
         keywords = dcatStrings.extract[String]('keywords.? / *),

